### PR TITLE
Initialize marsha plugins project

### PR DESCRIPTION
**Pull request by lunika**
## Purpose

To deliver marshs'a plugins we decided to host them on a AWS S3 bucket and distribute them using AWS CloudFront.
Adways is the first plugin we add to this repo.

## Proposal

- [x] create terraform configuration
- [x] add adways plugin

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+init: env.d/development
+	@bin/terraform init
+.PHONY: init
+
+deploy: ## Create or complete the AWS infra and copy files to the S3 bucket
+	@bin/terraform apply
+.PHONY: deploy
+
+help:  ## Show this help
+	@echo "$(BOLD)Marsha plugins Makefile$(RESET)"
```
Comment (jmaupetit) [inline Makefile:25]: Those variables are not set in your Makefile.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	@echo "Please use 'make $(BOLD)target$(RESET)' where $(BOLD)target$(RESET) is one of:"
```
Comment (jmaupetit) [inline Makefile:26]: Ditto

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	@grep -h ':\s\+##' Makefile | column -tn -s# | awk -F ":" '{ print "  $(BOLD)" $$1 "$(RESET)" $$2 }'
```
Comment (jmaupetit) [inline Makefile:27]: Ditto

File: .gitignore [hunk 4]
```diff
@@ -1 +1,4 @@
-env.d/development
\ No newline at end of file
+env.d/development
+.terraform
+*.tfstate
+*.tfstate.backup
```
Comment (jmaupetit) [inline .gitignore]: A carriage return is missing here.

File: env.d/development.dist [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# AWS admin credentials
+AWS_ACCESS_KEY_ID=yourAwsAccesKeyId
+AWS_SECRET_ACCESS_KEY=YourAwsSecretAccessKey
+
+TF_VAR_aws_region=eu-west-1
+AWS_DEFAULT_REGION=eu-west-1
```
Comment (jmaupetit) [inline env.d/development.dist]: Ditto.

File: terraform/cloudfront.tf [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  tags {
+    Environment = "${terraform.workspace}"
+  }
+
+  viewer_certificate {
+    cloudfront_default_certificate = true
+  }
+}
```
Comment (jmaupetit) [inline terraform/cloudfront.tf]: Ditto.

File: terraform/variables.tf [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  default = "eu-west-1"
+}
+
+variable "cloudfront_price_class" {
+  type = "map"
+
+  default = {
+    production = "PriceClass_All"
+  }
+}
```
Comment (jmaupetit) [inline terraform/variables.tf]: Ditto.

File: Readme.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# Marsha plugins
+
```
Comment (jmaupetit) [inline Readme.md:2]: This README should be commit along with the Makefile from whom rules are mentioned in.

File: Readme.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+This repo contains third party plugins we don't to commit in [marsha](https://github.com/openfun/marsha) application.
```
Comment (jmaupetit) [inline Readme.md]:
```suggestion
This repo contains optional plugins for the [Marsha](https://github.com/openfun/marsha) application.
```

File: Readme.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+plugins are pure javascript injected in the [lti template](https://github.com/openfun/marsha/blob/e5dec6b8ae662ab35950e5976296653cb29dbd7c/src/backend/marsha/core/templates/core/lti.html#L14-L16) and executed when the iframe is loaded.
```
Comment (jmaupetit) [inline Readme.md]:
```suggestion
Plugins are pure JavaScript files injected in the [LTI template](https://github.com/openfun/marsha/blob/e5dec6b8ae662ab35950e5976296653cb29dbd7c/src/backend/marsha/core/templates/core/lti.html#L14-L16) and executed when the target `iframe` is loaded.
```

File: Readme.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+Create the shared state bucket where `Terraform` will keep all the information on your deployments so different developers/machines/CI processes can interact with them:
+
+    $ make state-create
+
+Initialize your `Terraform` config:
+
+    $ make init
+
+In order to have multiple environment (development, staging, preproduction, production) you must create a terraform workspace and use it before deploying the project
```
Comment (jmaupetit) [inline Readme.md]:
```suggestion
In order to have multiple environments (_e.g._ `development`, `staging`, `preproduction` and `production`) you must create a Terraform workspace and use it before deploying the project:
```

File: Readme.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+    $ ./bin/terraform workspace create development
+
+Then you can create the architecture and deploy all third party plugins
```
Comment (jmaupetit) [inline Readme.md]:
```suggestion
Then you can create the architecture and deploy all plugins:
```

File: Readme.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+    $ make deploy
+
+At then end of the deployment terraform will output the AWS CloudFront url you will use to distribute javascript files.
+
+## Available plugins
+
+### Adways
+
+Adways is an advertising interactive videos. The plugins is in `src/static/js/adways.js`. To distribute it the URL will be: `https://yourCloudFrontDomain.com/static/js/adways.js`.
```
Comment (jmaupetit) [inline Readme.md]:
```suggestion
Adways is an advertising interactive videos. The plugin lives in `src/static/js/adways.js`. To distribute it the URL will be: `https://yourCloudFrontDomain.com/static/js/adways.js`.
```

Comment (jmaupetit) [inline Readme.md]: > Adways is an advertising interactive videos.

Should be rephrased.

Comment (lunika) [reply Readme.md]: It took it on the adways website

Comment (jmaupetit) [reply Readme.md]: A word is missing in this sentence. Maybe:

> Adways is an advertising videos solution.

WDYT?

Comment (lunika) [reply Readme.md]: we should probably keep `interactive` instead of `advertising` because we use it for its interactivity

Comment (lunika) [reply Readme.md]: Done

Comment (jmaupetit) [reply Readme.md]: True that. I've been too fast in answering.

**Review by jmaupetit — APPROVED** [review]

**Pull request merged**